File: node/packages/botas/src/schema/activity.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

/** Represents a user or bot account in a channel. */
export interface ChannelAccount {
  /** Unique identifier for the account. */
  id: string;
  /** Display name of the account. */
  name?: string;
  /** Azure Active Directory object ID. */
  aadObjectId?: string;
  /** Role of the entity (e.g. `"user"`, `"bot"`). */
  role?: string;
  /** Additional channel-specific properties. */
  properties?: Record<string, unknown>;
}

/** Teams-extended channel account with additional user properties. */
export interface TeamsChannelAccount extends ChannelAccount {
  /** User principal name (email). */
  userPrincipalName?: string;
  /** Email address. */
  email?: string;
}

/** Represents a conversation in a channel. */
export interface ConversationAccount {
  /** Unique identifier for the conversation. */
  id: string;
  /** Display name of the conversation. */
  name?: string;
  /** Whether the conversation contains more than two participants. */
  isGroup?: boolean;
  /** Type of conversation (e.g. `"personal"`, `"channel"`). */
  conversationType?: string;
  /** Tenant ID the conversation belongs to. */
  tenantId?: string;
  /** Additional channel-specific properties. */
  properties?: Record<string, unknown>;
}

/** An entity associated with an activity (e.g. a mention). */
export interface Entity {
  /** Entity type (e.g. `"mention"`, `"clientInfo"`). */
  type: string;
  [key: string]: unknown;
}

/** A file or card attachment on an activity. */
export interface Attachment {
  /** MIME type of the attachment content. */
  contentType: string;
  /** URL to retrieve the attachment content. */
  contentUrl?: string;
  /** Embedded content (e.g. an Adaptive Card JSON object). */
  content?: unknown;
  /** Filename or display name. */
  name?: string;
  /** URL to a thumbnail image. */
  thumbnailUrl?: string;
}

/** A reaction (emoji) applied to a message. */
export interface MessageReaction {
  /** Reaction type string (e.g. `"like"`, `"heart"`). */
  type: string;
}

/**
 * The base activity payload received from the Bot Framework.
 */
export interface Activity {
  /** Activity type (e.g. `"message"`, `"conversationUpdate"`). */
  type: string;
  /** Activity ID assigned by the channel. */
  id?: string;
  /** Channel identifier (e.g. `"msteams"`, `"directline"`). */
  channelId: string;
  /** Service URL for the sending channel — used to reply. */
  serviceUrl: string;
  /** Account that sent the activity. */
  from: ChannelAccount;
  /** Account that received the activity (the bot). */
  recipient: ChannelAccount;
  /** Conversation this activity belongs to. */
  conversation: ConversationAccount;
  /** Channel-specific payload (e.g. Teams channel data). */
  channelData?: unknown;
  /** Structured entities attached to the activity. */
  entities?: Entity[];
  /** File or card attachments. */
  attachments?: Attachment[];
  /** Activity value payload (used by invoke and event activities). */
  value?: unknown;
  /** ID of the activity this is a reply to. */
  replyToId?: string;
  /** Text content of a message activity. */
  text?: string;
  /** Name of the event or invoke operation. */
  name?: string;
  /** Action for installationUpdate activities (e.g. `"add"`, `"remove"`). */
  action?: string;
  /** Locale of the activity (e.g. `"en-US"`). */
  locale?: string;
  /** UTC timestamp when the activity was sent. */
  timestamp?: string;
  /** Reactions added to a message (messageReaction activities). */
  reactionsAdded?: MessageReaction[];
  /** Reactions removed from a message (messageReaction activities). */
  reactionsRemoved?: MessageReaction[];
  /** Members added to the conversation (conversationUpdate activities). */
  membersAdded?: ChannelAccount[];
  /** Members removed from the conversation (conversationUpdate activities). */
  membersRemoved?: ChannelAccount[];
  /** Additional custom properties. */
  properties?: Record<string, unknown>;
}

/** Response returned after creating or sending an activity. */
export interface ResourceResponse {
  /** ID of the created resource. */
  id: string;
}

/** Response returned after creating a new conversation. */
export interface ConversationResourceResponse extends ResourceResponse {
  /** Service URL for the new conversation. */
  serviceUrl: string;
  /** ID of the activity that bootstrapped the conversation. */
  activityId: string;
}

/** A page of members with an optional continuation token for subsequent pages. */
export interface PagedMembersResult<T> {
  /** Members on this page. */
  members: T[];
  /** Opaque token to pass on the next call to retrieve the following page. */
  continuationToken?: string;
}

/** Parameters used to create a new conversation. */
export interface ConversationParameters {
  /** Whether to create a group conversation. */
  isGroup?: boolean;
  /** Bot account to include in the conversation. */
  bot?: ChannelAccount;
  /** Initial member list. */
  members?: ChannelAccount[];
  /** Optional topic name. */
  topicName?: string;
  /** Tenant ID. */
  tenantId?: string;
  /** Initial activity to post when the conversation is created. */
  activity?: Partial<Activity>;
  /** Channel-specific data. */
  channelData?: unknown;
}

/** A page of conversations with an optional continuation token. */
export interface ConversationsResult {
  /** Conversations on this page. */
  conversations: ConversationAccount[];
  /** Opaque token to pass on the next call to retrieve the following page. */
  continuationToken?: string;
}

/** A transcript (ordered list of activities) for a conversation. */
export interface Transcript {
  /** Activities in the transcript, ordered by timestamp. */
  activities: Activity[];
}

/**
 * Create a reply activity from an incoming activity (FR-005).
 *
 * Copies `conversation`, `serviceUrl`, and `channelId` from the original,
 * swaps `from`/`recipient`, and sets `replyToId` to the original activity ID.
 *
 * @param activity - The incoming activity to reply to.
 * @param text - Optional text content for the reply.
 */
export function createReplyActivity (activity: Activity, text = ''): Partial<Activity> {
  return {
    type: 'message',
    channelId: activity.channelId,
    serviceUrl: activity.serviceUrl,
    conversation: activity.conversation,
    from: activity.recipient,
    recipient: activity.from,
    replyToId: activity.id,
    text,
  }
}

File: python/packages/botas/src/botas/schema/activity.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict, model_validator
from pydantic.alias_generators import to_camel


class _CamelModel(BaseModel):
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        extra="allow",
    )


class ChannelAccount(_CamelModel):
    id: str
    name: str | None = None
    aad_object_id: str | None = None
    role: str | None = None


class TeamsChannelAccount(ChannelAccount):
    user_principal_name: str | None = None
    email: str | None = None


class ConversationAccount(_CamelModel):
    id: str
    name: str | None = None
    is_group: bool | None = None
    conversation_type: str | None = None
    tenant_id: str | None = None


class Entity(_CamelModel):
    type: str


class Attachment(_CamelModel):
    content_type: str
    content_url: str | None = None
    content: Any = None
    name: str | None = None
    thumbnail_url: str | None = None


class MessageReaction(_CamelModel):
    type: str


class Activity(_CamelModel):
    type: str
    id: str | None = None
    channel_id: str = ""
    service_url: str = ""
    from_account: ChannelAccount | None = None
    recipient: ChannelAccount | None = None
    conversation: ConversationAccount | None = None
    channel_data: Any = None
    entities: list[Entity] | None = None
    attachments: list[Attachment] | None = None
    value: Any = None
    reply_to_id: str | None = None
    text: str | None = None
    name: str | None = None
    action: str | None = None
    locale: str | None = None
    timestamp: str | None = None
    reactions_added: list[MessageReaction] | None = None
    reactions_removed: list[MessageReaction] | None = None
    members_added: list[ChannelAccount] | None = None
    members_removed: list[ChannelAccount] | None = None

    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        extra="allow",
        # 'from' is a Python keyword — remapped via model_validator below
    )

    @model_validator(mode="before")
    @classmethod
    def _remap_from(cls, data: Any) -> Any:
        if isinstance(data, dict) and "from" in data:
            data = dict(data)
            data["from_account"] = data.pop("from")
        return data

    @classmethod
    def model_validate_json(cls, json_data: str | bytes, **kwargs: Any) -> "Activity":  # type: ignore[override]
        import json
        data = json.loads(json_data)
        return cls.model_validate(data, **kwargs)

    def model_dump(self, **kwargs: Any) -> dict[str, Any]:  # type: ignore[override]
        d = super().model_dump(**kwargs)
        # remap 'from_account' → 'from' in output
        if "from_account" in d:
            d["from"] = d.pop("from_account")
        elif "fromAccount" in d:
            d["from"] = d.pop("fromAccount")
        return d


class ResourceResponse(_CamelModel):
    id: str


class ConversationResourceResponse(ResourceResponse):
    service_url: str
    activity_id: str


class PagedMembersResult(_CamelModel):
    members: list[ChannelAccount] = []
    continuation_token: str | None = None


class ConversationParameters(_CamelModel):
    is_group: bool | None = None
    bot: ChannelAccount | None = None
    members: list[ChannelAccount] | None = None
    topic_name: str | None = None
    tenant_id: str | None = None
    activity: dict[str, Any] | None = None
    channel_data: Any = None


class ConversationsResult(_CamelModel):
    conversations: list[ConversationAccount] = []
    continuation_token: str | None = None


class Transcript(_CamelModel):
    activities: list[Activity] = []


def create_reply_activity(activity: Activity, text: str = "") -> dict[str, Any]:
    """Create a reply activity from an incoming activity (FR-005).

    Copies conversation, serviceUrl, and channelId from the original,
    swaps from/recipient, and sets replyToId to the original activity ID.
    Returns a plain camelCase dict ready for JSON serialization.
    """
    return {
        "type": "message",
        "channelId": activity.channel_id,
        "serviceUrl": activity.service_url,
        "conversation": (
            activity.conversation.model_dump(by_alias=True, exclude_none=True) if activity.conversation else None
        ),
        "from": activity.recipient.model_dump(by_alias=True, exclude_none=True) if activity.recipient else None,
        "recipient": (
            activity.from_account.model_dump(by_alias=True, exclude_none=True) if activity.from_account else None
        ),
        "replyToId": activity.id,
        "text": text,
    }
